### PR TITLE
Add MQTT-SN library

### PR DESCRIPTION
**Pull request by BeckmaR**
This library implements MQTT-SN version 1.2. It is transport agnostic;
users will have to implement a transport themselves. This make sense
since MQTT-SN does not require any kind of transport or protocol, it
just has to be frame-based.

The library uses a k_heap for dynamic memory allocation. This allows to
make use of the larger messages introduced in v1.2, without requiring
enormous amounts of memory to be pre-allocated.

This library has not been tested with an actual gateway. There are
however unit tests that make sure that package encoding and decoding
work correctly, as well as some basic connect and publish logic.

Signed-off-by: Rene Beckmann <rene.bckmnn@gmail.com>

---------
I have worked on this on and off for the last months. I would love to get some feedback on this - is it generally interesting, should I do something fundamentally different, does it make sense?

### CODE REVIEW

**Comment by carlescufi** [comment]
Thanks for the PR!
I am not an MQTT expert, but is there a reason why you introduce a new MQTT implementation instead of extending the existing one?

File: include/net/mqtt_sn.h
Comment (rlubos) [inline include/net/mqtt_sn.h:115]: I see that the approach here regarding communication (socket) handling is slightly different than in the core MQTT lib we have. Here, we expect the application to handle the communication part, while the standard MQTT lib handles sockets on its own, with an abstracted transport layer. I wonder if it made sense to follow that approach here, to give users unanimous feel of both libraries.

File: include/net/mqtt_sn.h
Comment (rlubos) [inline include/net/mqtt_sn.h:121]: Again, this is slightly different from our MQTT library, where we only register a single callback, which can handle different event types. I think we could do the same here, to remain consistent.

File: include/net/mqtt_sn.h
Comment (rlubos) [inline include/net/mqtt_sn.h:145]: Could use a more meaningful name here.

File: include/net/mqtt_sn.h
Comment (rlubos) [inline include/net/mqtt_sn.h:37]: Why do they need to be packed?

File: subsys/net/lib/mqtt_sn/Kconfig
Comment (rlubos) [inline subsys/net/lib/mqtt_sn/Kconfig:3]: Please update copyright (here and some other files)

**Review by rlubos — COMMENTED** [review]
Thank you for this contribution, I have to say it looks quite impressive. I've left a few inital high-level comments.

@carlescufi I see that the overall design of the library is different from the core MQTT library, but I think that it's somewhat justified with MQTT-SN:
1. The library encapsulates retransmission logic within, as MQTT-SN is built upon non-reliable UDP transport, I think it's the right choice to hande this within the library.
2. I was a bit reluctant initially about the heap usage in the library, but after a closer look I think it makes sense in this case - especially that it has a dedicated heap. First, we need memory to store the messages for reatransmission, so we can't just rely on stack here. Second, with UDP we can't send packets partially as we do in regular MQTT, we need to compose header/payload into a single linear buffer before transmission, due to the nature of UDP sockets.
3. And generally, while the general idea behind MQTT-SN is similar to MQTT, we're not interfacing MQTT Broker directly here, but rather an intermediate MQTT Gateway - the communication scheme between client/gateway is a bit different than between client/broker. I'm not really sure if extending the existing implementation to match SN requirements would be a viable choice here.

I would consider though to relieve the application from handling the sockets on it's own as we do in the regular MQTT lib. It's up to discussion though, if there are arguments not to do so I can reconsider.

File: include/net/mqtt_sn.h
Comment (jukkar) [reply include/net/mqtt_sn.h:37]: And if it really needs to be packed, there is the `__packed` define that can be used here.

File: include/net/mqtt_sn.h
Comment (BeckmaR) [reply include/net/mqtt_sn.h:115]: This was a design decision I made because MQTT-SN does not require any specific kind of transport, unlike MQTT which does require a reliable transport like TCP / TLS. You can use UDP for MQTT-SN, but you could also use ZigBee, UART, GATT, LoRaWAN or potentially even SMS.

It might make sense to take inspiration from the original MQTT library to have some kind of transport-related struct of function pointers, and provide a default one for UDP. That would make the approach similar, even if the details will differ.

File: include/net/mqtt_sn.h
Comment (BeckmaR) [reply include/net/mqtt_sn.h:121]: That might make sense, even though this library exposes fewer events than the original MQTT library. For example, I handle stuff like ACKs internally, where the MQTT library leaves that as an exercise to the user.

File: include/net/mqtt_sn.h
Comment (BeckmaR) [reply include/net/mqtt_sn.h:37]: It does make the parameter-union in the encoder and decoder parts much smaller. In an earlier version I had to store those for re-sends. Now, they are still used in the `mqtt_sn_publish` and `mqtt_sn_topic` structs. I thought using 3 bytes instead of 12 might be useful (lots of assumptions here).

File: subsys/net/lib/mqtt_sn/mqtt_sn.c
Comment (BeckmaR) [reply subsys/net/lib/mqtt_sn/mqtt_sn.c]: Done, hope I got all.

File: subsys/net/lib/mqtt_sn/Kconfig
Comment (BeckmaR) [reply subsys/net/lib/mqtt_sn/Kconfig]: Done.

**Comment by BeckmaR** [comment]
@carlescufi In the end, MQTT-SN and MQTT are different enough in some aspects that I thought it would really be more confusing to have them side-by-side in the same library than just build a completely separate one. Also there are some aspects about the original library that I don't like, e.g. that it does not handle ACKs and PUBREC etc itself.

File: include/net/mqtt_sn.h
Comment (BeckmaR) [reply include/net/mqtt_sn.h]: @rlubos The more I go into the details _how_ the MQTT library handles the transport, the less I like it, actually. The user is expected to call `mqtt_input` when there is new data. To know if that is the case, the user of course has to poll the socket. This breaks the separation of concerns IMO (see `zephyr/samples/net/cloud/google_iot_mqtt/src/protocol.c:215`, `wait_for_input`).

Comment (rlubos) [reply include/net/mqtt_sn.h]: Well, there was an argument that it should be possible to run MQTT along with other protocols on a single thread, therefore `mqtt_input()` is non-blocking and relies on the application to actually monitor the activity on a socket (so that it can add whatever sockets it wants to the `fds`). But perhaps it wouldn't be a bad idea to have some kind of `mqtt_wait()` function.

File: subsys/net/lib/mqtt_sn/mqtt_sn.c
Comment (rlubos) [inline subsys/net/lib/mqtt_sn/mqtt_sn.c:1024]: Could we drop the empty functions? I don't see a reason to introduce them, if particular replies don't require any processing.

File: include/zephyr/net/mqtt_sn.h
Comment (rlubos) [inline include/zephyr/net/mqtt_sn.h:236]: Is this function actually implemented anywhere?

File: subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c
Comment (rlubos) [inline subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:40]: The `recepient` field does not seem to be used, what's its purpose?

File: subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c
Comment (rlubos) [inline subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:26]: There should also be some function that closes the socket. In the regular `MQTT` implementation the approach is a bit different, the socket is created on "connect" and closed on "disconnect" calls, but it's a bit different due to TCP (the socket is no longer usable after TCP disconnect). Perhaps we need some kind of `deinit()` function here?

**Review by rlubos — COMMENTED** [review]
The library/test looks pretty good, I assume the sample part is still in early WIP?

**Comment by BeckmaR** [comment]
Hey @rlubos, thank you for the review. Due to multiple reasons I didn't get around to working on this in the meantime. I have now started to work on it again (fingers crossed) and changed some things that I did not like before:
- Removed the useless `struct mqtt_sn_msg` and everything related to it - the encoder and decoder now just use a `net_buf_simple`
- Removed the terrible heap, I aligned with libraries like lwm2m and mqtt and simply introduced n times x bytes of memory (globally) for publish data and topics
- Removed the gateway void pointer. I decided not to support broadcasting and SEARCHGW currently, users will just have to provide e.g. a gateway IP and port to the UDP transport, or take care of that somewhere else when using e.g. Bluetooth

I will now take your other comments into account and e.g. remove the empty functions. Next step would be to create a sample application and test this against an actual gateway.

File: include/zephyr/net/mqtt_sn.h
Comment (BeckmaR) [reply include/zephyr/net/mqtt_sn.h:208]: no, removed this

File: subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c
Comment (BeckmaR) [reply subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:40]: removed

File: include/zephyr/net/mqtt_sn.h
Comment (BeckmaR) [reply include/zephyr/net/mqtt_sn.h:198]: renamed to `process_work`

File: include/net/mqtt_sn.h
Comment (BeckmaR) [reply include/net/mqtt_sn.h]: removed these

**Review by rlubos — COMMENTED** [review]
This looks quite neat, thank you for the work you put into this library. I have a few final requests:
* Please split the library/sample/tests into separate commits, each commit following the [commit guidelines](https://docs.zephyrproject.org/3.2.0/contribute/guidelines.html#commit-guidelines) (title/body/signoff)
* There are quite a few style issues, please check https://github.com/zephyrproject-rtos/zephyr/actions/runs/3329890091/jobs/5507679599 for details.
* Looks that some minor fixes are needed in the documentation as well: https://github.com/zephyrproject-rtos/zephyr/actions/runs/3329890086/jobs/5507679754#step:7:51
* Please add `sample.yaml` to the sample, so that it'll be built by the CI (here's an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/mqtt_publisher/sample.yaml)
* And finally, please add an entry for the library in the [CODEOWNERS](https://github.com/zephyrproject-rtos/zephyr/blob/main/CODEOWNERS) and [MAINTAINERS](https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml) file.

File: samples/net/mqtt_sn_publisher/src/common.h
Comment (BeckmaR) [reply samples/net/mqtt_sn_publisher/src/common.h:2]: That is intentional since I did some parts on company time. Thank you for watching out!

File: tests/net/lib/mqtt_sn_client/prj.conf
Comment (BeckmaR) [reply tests/net/lib/mqtt_sn_client/prj.conf:13]: Ah, it was a crudge to compile the library with these options in the beginning - leftover.

**Comment by BeckmaR** [comment]
@rlubos Still polishing some stuff, some tests fail due to minor errors. If everything builds I will split it up into smaller commits.

I can't figure out why the documentation build fails, it complains it can't find LaTeX - I didn't touch anything in that area.

**Comment by BeckmaR** [comment]
@rlubos CI is happy, as mentioned above I have no idea about what you expect for the CODEOWNERS and MAINTAINERS files.

**Comment by rlubos** [comment]
> @rlubos CI is happy, as mentioned above I have no idea about what you expect for the CODEOWNERS and MAINTAINERS files.

The changes in those files are just what I expected, thanks. @carlescufi Is the GH warning from the CODEOWNERs file something we should be concerned about? I see that there are already entries like the one added, where not all users have write access to the repo.

**Review by rlubos — DISMISSED** [review]
LGTM, thank you for the PR, I have no more concerns.

**Comment by stephanosio** [comment]
> The changes in those files are just what I expected, thanks. @carlescufi Is the GH warning from the CODEOWNERs file something we should be concerned about? I see that there are already entries like the one added, where not all users have write access to the repo.

No, since the user has been added to the MAINTAINERS file (which is the authoritative one) as a collaborator, the user will be granted the collaborator access once this PR is merged and the warning will no longer apply.

File: samples/net/mqtt_sn_publisher/Kconfig
Comment (jukkar) [inline samples/net/mqtt_sn_publisher/Kconfig:14]: For the sample apps, I would suggest that `SAMPLE` is prefixed to config symbols so `SAMPLE_MQTT_SN_GATEWAY_IP` etc. This makes it clear that the symbol is only available for this sample. See some other network samples for examples how to do it.

File: subsys/net/lib/mqtt_sn/Kconfig
Comment (rlubos) [reply subsys/net/lib/mqtt_sn/Kconfig:7]: That's a good idea, we could keep the library experimental for a few releases so that we have more freedom finetuning the API if needed. @BeckmaR Here's an example on how to set a library as experimental: https://github.com/zephyrproject-rtos/zephyr/blob/f483b1bc4cfe2381624fac6337cc74d3ce2bc850/subsys/net/lib/tftp/Kconfig#L6 (note the `select EXPERIMENTAL` is also needed).

File: samples/net/mqtt_sn_publisher/Kconfig
Comment (BeckmaR) [reply samples/net/mqtt_sn_publisher/Kconfig]: Good idea - done.

File: subsys/net/lib/mqtt_sn/Kconfig
Comment (BeckmaR) [reply subsys/net/lib/mqtt_sn/Kconfig]: Done that, makes sense. Thanks for the input.

**Review by jukkar — DISMISSED** [review]
LGTM

**Review by rlubos — DISMISSED** [review]

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**
